fv3core/stencils/delnflux.py

```diff
@@ -1,218 +1,627 @@
+            fx2: diffusive x-flux on A grid (in/out)
+            fy2: diffusive y-flux on A grid (in/out)
+            damp_c: damping coefficient for q (in)
+            d2: A damped copy of the q field (in)
+            mass: Mass to weight the diffusive flux by (in)
+        """
+        grid = spec.grid
+
+        if mass is None:
+            self._d2_damp(q, d2, damp_c)
+        else:
+            self._copy_stencil_interval(q, d2)
+
+        for kstart, k_range in enumerate(self._k_bounds):
+            if self._nord[kstart] > 0:
+                corners.copy_corners_x_stencil(
+                    d2,
+                    origin=(grid.isd, grid.jsd, kstart),
+                    domain=(grid.nid, grid.njd, k_range),
+                )
+
+        self._fx_calc_stencil(d2, grid.del6_v, fx2)
+
+        for kstart, k_range in enumerate(self._k_bounds):
+            if self._nord[kstart] > 0:
+                corners.copy_corners_y_stencil(
+                    d2,
+                    origin=(grid.isd, grid.jsd, kstart),
+                    domain=(grid.nid, grid.njd, k_range),
+                )
+
+        self._fy_calc_stencil(d2, grid.del6_u, fy2)
+
+        for n in range(self._nmax):
+            nt = self._nmax - 1 - n
+            nt_origin = (grid.is_ - nt - 1, grid.js - nt - 1, 0)
             nt_ny = grid.je - grid.js + 3 + 2 * nt
             nt_nx = grid.ie - grid.is_ + 3 + 2 * nt
-            d2_highorder(
-                fx2, fy2, grid.rarea, d2, origin=nt_origin, domain=(nt_nx, nt_ny, nk)
-            )
-            corners.copy_corners_x_stencil(
-                d2, origin=(grid.isd, grid.jsd, kstart), domain=(grid.nid, grid.njd, nk)
+            self._d2_stencil(
+                fx2,
+                fy2,
+                grid.rarea,
+                d2,
+                origin=nt_origin,
+                domain=(nt_nx, nt_ny, self._nk),
             )
-            nt_origin = (grid.is_ - nt, grid.js - nt, kstart)
-            fx2_order(
+
+            for kstart, k_range in enumerate(self._k_bounds):
+                if self._nord[kstart] > 0:
+                    corners.copy_corners_x_stencil(
+                        d2,
+                        origin=(grid.isd, grid.jsd, kstart),
+                        domain=(grid.nid, grid.njd, k_range),
+                    )
+
+            nt_origin = (grid.is_ - nt, grid.js - nt, 0)
+            self._column_conditional_fx_calculation(
                 d2,
                 grid.del6_v,
                 fx2,
-                order=2 + n,
                 origin=nt_origin,
-                domain=(nt_nx - 1, nt_ny - 2, nk),
-            )
-            corners.copy_corners_y_stencil(
-                d2, origin=(grid.isd, grid.jsd, kstart), domain=(grid.nid, grid.njd, nk)
+                domain=(nt_nx - 1, nt_ny - 2, self._nk),
             )
 
-            fy2_order(
+            for kstart, k_range in enumerate(self._k_bounds):
```

Comment:
maybe I am misunderstanding something, but can't we just apply corners on the full 3d domain? this seems to do that but more convoluted?

Reply:
Nord is 0 for some k-levels, and the corners shouldn't get copied for those k-levels